### PR TITLE
types/views: add JSON marshal/unmarshal and AsMap to Map

### PR DESCRIPTION
**Pull request by maisem**
This allows cloning a Map as well as marshaling the Map as JSON.

Updates tailscale/corp#12754

### CODE REVIEW

File: types/views/views.go [hunk 1]
```diff
@@ -332,6 +333,26 @@ func (m Map[K, V]) GetOk(k K) (V, bool) {
 	return v, ok
 }
 
+// MarshalJSON implements json.Marshaler.
+func (m Map[K, V]) MarshalJSON() ([]byte, error) {
+	return json.Marshal(m.ж)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (m *Map[K, V]) UnmarshalJSON(b []byte) error {
```
Comment (dsnet) [inline types/views/views.go:342]: The whole point of a view is to provide immutable access to the underlying type. Wouldn't `UnmarshalJSON` violate this goal?

Comment (bradfitz) [reply types/views/views.go:342]: See Maisem's latest push.

Comment (maisem) [reply types/views/views.go:342]: Fixed with an additional check. This is how we handle it in other places.
```
	if v.ж != nil {
		return errors.New("already initialized")
	}
```

**Comment by bradfitz** [comment]
If you need to clone, add a clone?

If you need to a view of some JSON, unmarshal it into T and then make a View[T] ?

File: types/views/views.go (hunk 1 above)
Comment (dsnet) [reply types/views/views.go:342]: Should this use the `unmarshalJSON` helper method like most other code here?

File: types/views/views.go [hunk 2]
```diff
@@ -332,6 +333,29 @@ func (m Map[K, V]) GetOk(k K) (V, bool) {
 	return v, ok
 }
 
+// MarshalJSON implements json.Marshaler.
+func (m Map[K, V]) MarshalJSON() ([]byte, error) {
+	return json.Marshal(m.ж)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
```
Comment (bradfitz) [inline types/views/views.go:341]: say that it only works on the zero value?

File: types/views/views.go [hunk 3]
```diff
@@ -332,6 +333,29 @@ func (m Map[K, V]) GetOk(k K) (V, bool) {
...
+func (m *Map[K, V]) UnmarshalJSON(b []byte) error {
+	if m.ж != nil {
+		return errors.New("already initialized")
+	}
+	return json.Unmarshal(b, &m.ж)
+}
+
+// AsMap returns a deep-clone of the underlying map.
```
Comment (bradfitz) [inline types/views/views.go:349]: shallow clone, no?

File: types/views/views.go (hunk 1 above)
Comment (maisem) [reply types/views/views.go:343]: >Should this use the unmarshalJSON helper method like most other code here?

that only works for slices, I didn't spend too much time make it work on both slices and maps

**Review by bradfitz — APPROVED** [review]

**Review by dsnet — APPROVED** [review]

**Pull request merged**